### DISCUSSION TITLE
JupyterLab Cocotb extension

### DISCUSSION THREAD

**Discussion by mmichilot** [cocotb/cocotb#3575] [Cocotb In The Wild]
This is still in early development, but I'm building a JupyterLab extension that adds support for Cocotb. Here's a small demo of it in action!
![CocotbDemo](https://github.com/cocotb/cocotb/assets/22762462/8afe8d60-a081-4bda-b475-fd9a6a699b00)

The extension uses a custom kernel provisioner to build and launch a design using cocotb's new test runners. Then the `ipykernel` is launched once cocotb is initialized. 

To setup a design, I also developed a frontend:
![Screenshot 2023-12-10 183000](https://github.com/cocotb/cocotb/assets/22762462/29362a64-607d-433b-ad68-a875700fcdfd)

I haven't made the extension public since I'm waiting for JupyterLab to merge some of my changes, but I'll keep this discussion updated.

**Comment by cmarqu** [comment]
Wow, this looks great!

**Comment by mmichilot** [comment]
Here's another demo testing a D Flip Flop with waveform output:
![DFlipFlopDemo](https://github.com/cocotb/cocotb/assets/22762462/ccbea904-94cd-4ad1-9940-c8bd81be26c1)

**Reply by ktbarrett** [reply]
When it's ready let us know so we can add it to the list of interesting related projects on our wiki.

**Comment by cmarqu** [comment]
Hey @mmichilot, any news yet?

**Comment by boehmerst** [comment]
Hey @mmichilot, I was just searching for exactly such a solution. Is there any news?

**Comment by mmichilot** [comment]
Hey @cmarqu and @boehmerst, apologies for the long delay and lack of updates. I've decided to take a different approach and create a  Jupyter kernel rather than an JupyterLab extension. I'm excited to finally release the kernel, [cocotb_kernel](https://github.com/mmichilot/cocotb_kernel). 

I've also created a new discussion, #3981, since this is no longer a JupyterLab extension.
